streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -529,7 +528,7 @@ private void checkForException() {
         final KafkaException exception = sendException.get();
 
         if (exception != null) {
-            sendException.set(null);
+            sendException.compareAndSet(exception, null);
```

Comment:
I believe this is actually fixing a race condition... with `set()` we might overwrite (and thus drop) and exception which is set after we called `get()` (even if I am not 100% sure how bad it is if we drop it, give that we `throw exception;` below anyway...?

Reply:
Yeah... not sure that part makes a difference. But I guess there is nothing wrong with it?

The rest of PR makes sense to me